File: Algorithms/Longest Substring Without Repeating Characters/longest-substring-without-repeating-characters.js
// Source : https://leetcode.com/problems/longest-substring-without-repeating-characters/
// Author : Han Zichi
// Date   : 2016-08-05

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
  var hash = {};
  var start = 0;
  var ans = 0;

  for (var i = 0, len = s.length; i < len; i++) {
    var item = s[i];

    if (!hash[item])
      hash[item] = true;
    else {
      // item 已经在 substring 中存在了
      for (; ;) {
        if (s[start] === item) {
          start++;
          break;
        }

        hash[s[start]] = false;
        start++;
      }
    }

    ans = Math.max(ans, i - start + 1);
  }

  return ans;
};

File: Algorithms/Longest Substring Without Repeating Characters/longest-substring-without-repeating-characters.py
#!/usr/bin/env python
# coding=utf-8


# Source : https://leetcode.com/problems/longest-substring-without-repeating-characters/
# Author : Meow-z
# Date   : 2017-01-04


# 122ms
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        new_list = []
        max_list = []
        for i in s:
            if len(new_list) >= len(max_list):
                max_list = new_list
            if i in new_list:
                list_index = new_list.index(i)
                new_list = new_list[list_index+1:]
            new_list.append(i)
        return len(max_list)